batch-deletion/main.py

```diff
@@ -0,0 +1,167 @@
+
+        dir_count = 0
+        file_count = 0
+        total_size = 0
+        if queue_item.Type == "Block":
+            blob_list = azure.list_storage_blobs(queue_item.Container, queue_item.ListPath)
+            blob_path_list = []
+            for blob in blob_list:
+                blob_info = azure.get_blob_info(blob)
+                if blob_info["dir"]:
+                    dir_count += 1
+                    PendingList.put(QueueItem(Container=queue_item.Container, Type="Block", ListPath=blob_info["name"]))
+                else:
+                    file_count += 1
+                    total_size += blob_info["size"]
+                    blob_path_list.append(blob_info["name"])
+                blob_info.clear()
+
+            if len(blob_path_list) > 0:
+                UsageSummary.put(DeletionQueue(Container=queue_item.Container, BlobList=blob_path_list))
+
+        PendingList.task_done()
+
+        IteratorLock.acquire()
+        IterationRunning -= 1
+        IteratorLock.release()
+
+    # print("Child " + str(thrId) + " exited : Dir " + str(total_path_processed))
+    IteratorLock.acquire()
+    MaxThreadCount -= 1
+    IteratorLock.release()
+
+
+
+def delete_batch():
+    Done = False
+    while not Done:
+        try:
+            usage_item = UsageSummary.get(block=False, timeout=3)
+        except:
+            if MaxThreadCount == 0:
+                Done = True
+            continue
+
+        print(usage_item.BlobList)
+        azure.delete_blob_batch(usage_item.Container, usage_item.BlobList)
+        UsageSummary.task_done()
+
+
+# ----------------------- Main processing --------------------------
+def main(cli_options):
+    global MaxThreadCount
+
+    if cli_options['connection_string']:
+        azure.set_connection_string(cli_options['connection_string'])
+    else:
+        conn_str = os.getenv("AZURE_STORAGE_CONNECTION_STRING")
+        azure.set_connection_string(conn_str)
+
+    if cli_options['parallel_factor']:
+        MaxThreadCount = int(cli_options['parallel_factor'])
+
+    # Get the list of containers from the storage account
+    # Push them to queue for first level of iteration
+    global TotalContainer
+    container_list = azure.list_storage_containers()
+    TotalContainer = len(container_list)
+
+    if TotalContainer == 0:
+        print("There is nothing to iterate in this account")
+        exit()
+
+    for container in container_list:
+        PendingList.put(QueueItem(Container=container, Type="Block", ListPath=""))
+    # container_list.clear()
+
+    # Start the thread to report the usage info
+    report_thread = threading.Thread(target=delete_batch)
+    report_thread.daemon = True
+    report_thread.start()
+
+    # Start N threads to iterate the directory and list recursively
+    for i in range(MaxThreadCount):
+        # print("Starting child thread " + str(i))
+        t = threading.Thread(target=thread_handler, args=[i])
+        t.daemon = True
```

Comment:
Is there a reason why you're instantiating the threads as daemon threads? You're joining them later on - so they're technically acting as non-daemon threads.

Reply:
Ohh we didn't know about that. Never used Python before this hackathon actually :)